src/utility.cpp

```diff
@@ -423,38 +423,81 @@ BOOL IsModulePatched (HMODULE importmodule, moduleentry_t patchtable [], UINT ta
 
     // No patches listed in the patch table were found in the import module.
     return FALSE;
-}
-
+}
+
 LPVOID FindRealCode(LPVOID pCode)
-{
-    LPVOID result = pCode;
+{
+    LPVOID result;
     if (pCode != NULL)
     {
-        if (*(WORD *)pCode == 0x25ff) // JMP r/m32
-        {
+        // we need to make sure we can read the first 3 ULONG_PTRs
+        DWORD old_protect;
+        // make sure we can read the first 3 pointers
+        if (VirtualProtect(pCode, sizeof(ULONG_PTR) * 3, PAGE_EXECUTE_READ, &old_protect))
+        {
+            if (*(WORD*)pCode == 0x25ff) // JMP r/m32
+            {
 #ifdef _WIN64
-            LONG offset = *((LONG *)((ULONG_PTR)pCode + 2));
-            // RIP relative addressing
-            PBYTE pNextInst = (PBYTE)((ULONG_PTR)pCode + 6);
-            pCode = *(LPVOID*)(pNextInst + offset);
-            return pCode;
+                LONG offset = *((LONG*)((ULONG_PTR)pCode + 2));
+                // RIP relative addressing
+                PBYTE pNextInst = (PBYTE)((ULONG_PTR)pCode + 6);
+
+                // now that we got the offset, make sure we can read the code at the offset
+                DWORD old_protect_2;
+                if (VirtualProtect(pCode, sizeof(LPVOID), PAGE_EXECUTE_READ, &old_protect_2))

```

Comment:
>pCode [](start = 35, length = 5)

pNextInst + offset, here and below (or throw it in a variable for better readability)

Reply:
Fixed.

---
In reply to: [591983750](https://github.com/Azure/vld/pull/8#discussion_r591983750) [](ancestors = 591983750)